File: lv1_module2_minsikim/ros2_ws/src/turtle_cpp/src/distance_publisher.cpp
#include <chrono>
#include <cmath>
#include <functional>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/float32.hpp"
#include "turtlesim/msg/pose.hpp"

class DistancePublisher : public rclcpp::Node
{
public:
    DistancePublisher() : Node("distance_publisher")
    {
        pose_subscription_ = this->create_subscription<turtlesim::msg::Pose>(
            "/turtle1/pose",
            10,
            [this](const turtlesim::msg::Pose::SharedPtr msg)
            {
                this->pose_callback(msg);
            }
        );

        distance_publisher_ =
            this->create_publisher<std_msgs::msg::Float32>(
                "/turtle_distance",
                10
            );

        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(100),
            [this]()
            {
                this->publish_distance();
            }
        );
    }

private:
    void pose_callback(
        const turtlesim::msg::Pose::SharedPtr msg)
    {
        latest_x_ = msg->x;
        latest_y_ = msg->y;
    }

    void publish_distance()
    {
        double distance =
            std::sqrt(
                latest_x_ * latest_x_ +
                latest_y_ * latest_y_
            );

        std_msgs::msg::Float32 msg;
        msg.data = static_cast<float>(distance);

        distance_publisher_->publish(msg);
    }

    double latest_x_ = 0.0;
    double latest_y_ = 0.0;

    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr
        pose_subscription_;

    rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr
        distance_publisher_;

    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<DistancePublisher>();

    rclcpp::spin(node);

    rclcpp::shutdown();

    return 0;
}

File: lv1_module2_minsikim/ros2_ws/src/turtle_py/turtle_py/distance_publisher.py
import math

import rclpy
from rclpy.node import Node

from turtlesim.msg import Pose
from std_msgs.msg import Float32

class DistancePublisher(Node):
    def __init__(self):
        super().__init__('distance_publisher')
        
        self.latest_x = 0.0
        self.latest_y = 0.0
        
        self.pose_subscription = self.create_subscription(
            Pose,
            '/turtle1/pose',
            self.pose_callback,
            10
        )
        self.distance_publisher = self.create_publisher(
            Float32,
            '/turtle_distance',
            10
        )
        
        self.declare_parameter('publish_rate', 10.0)
        publish_rate = self.get_parameter('publish_rate').value
        period = 1.0 / publish_rate
        self.timer = self.create_timer(
            period,
            self.publish_distance
        )
        
    def pose_callback(self, msg):
        self.latest_x = msg.x
        self.latest_y = msg.y
        
    def publish_distance(self):
        distance = math.sqrt(self.latest_x ** 2 + self.latest_y ** 2)
        msg = Float32()
        msg.data = float(distance)
        
        self.distance_publisher.publish(msg)
        
        
def main(args=None):
    rclpy.init(args=args)

    node = DistancePublisher()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()